modules/images/regenerate-existing-images/background-process/class-perflab-background-process.php

```diff
@@ -0,0 +1,236 @@
+	public function __construct() {
+		add_action( 'wp_ajax_' . Perflab_Background_Process::BG_PROCESS_ACTION, array( $this, 'handle_request' ) );
+		add_action( 'wp_ajax_nopriv_' . Perflab_Background_Process::BG_PROCESS_ACTION, array( $this, 'handle_request' ) );
+	}
+
+	/**
+	 * Handle incoming request to run the batch of job.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @throws Exception Invalid request for background process.
+	 *
+	 * @return void
+	 */
+	public function handle_request() {
+		try {
+			// Exit if nonce varification fails.
+			$nonce_check = check_ajax_referer( self::BG_PROCESS_ACTION, 'nonce', false );
+
+			// If nonce check fails, fallback to key checking.
+			if ( false === $nonce_check ) {
+				throw new Exception( __( 'Invalid nonce passed to process.', 'performance-lab' ) );
+			}
+
+			$job_id = isset( $_REQUEST['job_id'] ) ? absint( sanitize_text_field( $_REQUEST['job_id'] ) ) : 0;
+
+			// Job ID is mandatory to be specified.
+			if ( empty( $job_id ) ) {
+				throw new Exception( __( 'No job specified to execute.', 'performance-lab' ) );
+			}
+
+			$this->job = new Perflab_Background_Job( $job_id );
+
+			// Silently exit if the job is not ready to run.
+			if ( ! $this->job->should_run() ) {
+				return;
+			}
+
+			$this->job->lock(); // Lock the process for this job before running.
+			$this->run(); // Run the job.
+
+			// Once job ran successfully, change its status to queued.
+			$this->job->set_status( Perflab_Background_Job::JOB_STATUS_PARTIAL );
+
+		} catch ( Exception $e ) {
+			if ( $this->job instanceof Perflab_Background_Job ) {
+				$error = new WP_Error( 'perflab_job_failure', $e->getMessage() );
+				$this->job->set_error( $error );
+			}
+		} finally {
+			if ( $this->job instanceof Perflab_Background_Job ) {
+				// Unlock the process once everything is done.
+				$this->job->unlock();
+				$this->next_batch();
+			}
+		}
+	}
+
+	/**
+	 * Runs the process over a batch of job.
+	 *
+	 * As of now, it won't fetch the next batch if memory or time
+	 * is not exceeded, but this can be introduced later.
+	 *
+	 * Consumer code can fine-tune the batch size according to requirement.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return void
+	 */
+	private function run() {
+		do {
+			/**
+			 * Consumer code will hook to this action to perform necessary tasks.
+			 *
+			 * @param array $data Job data.
+			 */
+			do_action( 'perflab_job_' . $this->job->get_name(), $this->job->get_data() );
+		} while ( ! $this->memory_exceeded() && ! $this->time_exceeded() );
+	}
+
+	/**
+	 * Call the next batch of the job if it is not completed already.
+	 *
+	 * This will send a POST request to admin-ajax.php with background
+	 * process specific action to continue executing the job in a new process.
+	 *
+	 * @return void
+	 */
+	private function next_batch() {
+		/**
+		 * Do not call the background process from within the script if the
+		 * real cron has been setup to do so.
+		 */
+		if ( defined( 'ENABLE_BG_PROCESS_CRON' ) ) {
```

Comment:
@ankitrox I think the configuration we have defined so far is to disable this background process entirely with a `DISABLE_BACKGROUND_PROCESS` constant. I don't believe we have the infrastructure defined to have the WP Cron handle the additional batches.

I think we should remove this here and consider where the `DISABLE_BACKGROUND_PROCESS` check would need to be made.

Reply:
We will also have this constant to let the real cron job call background process. In such cases, script would not call next batch of the job from within itself.
This is discussed with @felixarntz in alternate approaches to improve background process workflow.

Reply:
Thanks @ankitrox I'm not sure how this works in the grand scheme of things as we haven't defined or built the part that handles subsequent batches. I think we can approach this change in 1 of 2 ways.

1. Keep the constant as part of this PR, but make a change as this is used twice in the code, once here and again inside the `perflab_dispatch_background_process_request`. I think we should aim to only have to check this constant once. 

2. Remove the constant from the PR and we can add it as part of a new PR and stream of work to introduce this functionality.

I think in both cases we should also create a new GitHub issue with acceptance criteria to outline the work.

cc @felixarntz